### PR TITLE
Shade external dependencies in the rest client jar

### LINKED ISSUE #25208: Shade client jars and build uber jar
We currently have a shared version of http client and its dependencies (eg various apache commons libs). This version must be used across any plugin/module that uses the ES test framework. Note that we have no such dependency in core, yet still the version must alight because the test framework is pulled in everywhere, and the rest client is part of the test framework.  This is particularly troublesome as we sometimes force a version of these dependencies which does not align with what the plugin requires from other dependencies (eg see discussion in #24067).

In order to allow each plugin to use their own version of httpclient and its deps, we should shade http client and all other deps of the rest client, and package it in an uber jar. This will make the rest client standalone, so that no conflict will occur when the client is on the test classpath with any plugin also depending on http client.

### PR DESCRIPTION
**Pull request by hub-cap**
This commit removes all external dependencies from the rest client jar,
and shades them in an 'internal.' package within the jar, using
shadowJar gradle plugin. All projects that depended on the existing jar
have been converted to using the 'internal.' package prefixes to
interact with the rest client.

Closes #25208

### CODE REVIEW

File: client/rest/build.gradle [hunk 1]
```diff
@@ -38,13 +42,50 @@ publishing {
...
+  configurations = [project.configurations.shade]
+  classifier = 'deps'
+  relocate 'org.apache', 'internal.org.apache'
+
+  doLast {
+    shadedSrcDir.mkdir()
+  }
+}
+
+// Actual task for
```
Comment (hub-cap) [inline client/rest/build.gradle]: heh, will fix this comment

File: client/test/build.gradle [hunk 2]
```diff
@@ -27,7 +26,7 @@ targetCompatibility = JavaVersion.VERSION_1_7
 sourceCompatibility = JavaVersion.VERSION_1_7
 
 dependencies {
-  compile "org.apache.httpcomponents:httpcore:${versions.httpcore}"
+  compile "org.elasticsearch.client:elasticsearch-rest-client:${version}"
```
Comment (javanna) [inline client/test/build.gradle]: why is httpcore replaced with the low-level rest client here?

File: core/src/main/resources/org/elasticsearch/bootstrap/test-framework.policy [hunk 3]
```diff
@@ -69,6 +69,16 @@ grant codeBase "${codebase.elasticsearch-rest-client-6.0.0-beta1-SNAPSHOT.jar}"
...
+grant codeBase "${codebase.elasticsearch-rest-client-6.0.0-beta1-SNAPSHOT-nodeps.jar}" {
+  // rest makes socket connections for rest tests
+  permission java.net.SocketPermission "*", "connect";
+};
+
+grant codeBase "${codebase.elasticsearch-rest-client-6.0.0-beta1-SNAPSHOT-deps.jar}" {
+  // rest makes socket connections for rest tests
+  permission java.net.SocketPermission "*", "connect";
+};
+
```
Comment (javanna) [inline core/src/main/resources/org/elasticsearch/bootstrap/test-framework.policy]: don't we need to remove the existing jar entry without the suffix?

File: client/test/build.gradle (hunk 2 above)
Comment (javanna) [inline client/test/build.gradle]: I think I get it. either we use the ordinary httpcore classes, or the internal ones but they come from our own big jar.

**Review by javanna — COMMENTED** [review]
I left a couple of comments, I will leave the gradle part to @rjernst for review. Did I get it right that the plan is to have two artifacts, one with deps and the other one without?

**Comment by hub-cap** [comment]
@javanna Ill leave some better comments in the github commit on why the artifacts are split out like that. its somewhat confusing.

File: client/test/build.gradle (hunk 2 above)
Comment (hub-cap) [inline client/test/build.gradle]: well, we really cant use ordinary ones now. Since all of the dependencies in rest client are shaded to use internal.org.apache classes, if we leak anything like `Header` into an API, we need to use the internal.org.apache since those are in the method signatures. So, we have to use our internal ones always, which is why we removed the httpcore stuff, because client/test now uses `internal.o.a.h.Header`, for example.

File: core/src/main/resources/org/elasticsearch/bootstrap/test-framework.policy (hunk 3 above)
Comment (hub-cap) [inline core/src/main/resources/org/elasticsearch/bootstrap/test-framework.policy]: Hopefully this will be explained better in the commit msg update

**Comment by hub-cap** [comment]
ok I decided to just add better docs to the build.gradle instead. Its saner there anyway.

File: client/rest/build.gradle [hunk 4]
```diff
@@ -38,13 +55,50 @@ publishing {
...
+  doLast {
+    shadedSrcDir.mkdir()
+  }
+}
+
+// builds the actual shaded jar that is being distributed
+shadowJar {
+  configurations = [project.configurations.shade]
+  classifier = null
+  relocate 'org.apache', 'internal.org.apache'
```
Comment (rjernst) [inline client/rest/build.gradle]: Shouldn't there be no need to use `relocate` here? If there is then unit tests ran with unshaded classes, which is exactly what we were trying to avoid.

File: core/src/main/resources/org/elasticsearch/bootstrap/test-framework.policy (hunk 3 above)
Comment (rjernst) [inline core/src/main/resources/org/elasticsearch/bootstrap/test-framework.policy]: I also do not understand why both of these are necessary. The rest client used by the test framework for REST integ tests should be using the jar with no classifier. The deps version above should replace the 2 http grants below? I don't see where the nodeps version would come in.

**Review by rjernst — COMMENTED** [review]
I left a couple comments. I also think we should think of a better package name prefix, since these are part of the api (sorry I did not realize that when suggesting the name in the first place!).

File: client/rest/build.gradle [hunk 5]
```diff
@@ -19,6 +19,19 @@
...
+ *
+ * 1) The original jar contents, using the `internal.` packages, but without the `internal.` code also in the jar.
+ *    The reason this jar was not overwritten was because gradle knows what its inputs and outputs are for a given task
+ *    and we should not alter that by overwriting the original jar. What we do is add a `nodeps` classifier to it instead.
+ * 2) A jar that contains *only* the `internal.` dependencies. This is a jar that is built before the src is compiled.
+ *    This jar is only used by the rest client so its compilable. There exists a chicken-egg situation where the src
+ *    needs compilation and depends on `internal.` classes, so an intermediary jar needs to exist to satisfy the compile.
+ *    The `deps` classifier is added to this jar.
+ * 3) The *actual* jar that will be used by clients. This has no classifier, contains the rest client src and `internal.`.
+ */
```
Comment (javanna) [inline client/rest/build.gradle]: thanks for adding this comment!

File: client/test/build.gradle (hunk 2 above)
Comment (javanna) [inline client/test/build.gradle]: thanks for the explanation!

**Comment by hub-cap** [comment]
@javanna @rjernst I have cleaned up the build so that it uses the actual shaded jar outside of the rest project, instead of relying on the transient project dependencies (which was the -deps.jar). I have also removed the -nodeps jar from existence outside of the build, so the shaded jar is the only thing that other projects know about / build to.

File: client/rest/build.gradle (hunk 4 above)
Comment (hub-cap) [inline client/rest/build.gradle]: This is the duplicated shadowing task that actually shades the classes in src as well as the `internal.` things.

File: core/src/main/resources/org/elasticsearch/bootstrap/test-framework.policy (hunk 3 above)
Comment (hub-cap) [inline core/src/main/resources/org/elasticsearch/bootstrap/test-framework.policy]: It doesn't, anymore :)

File: client/rest/build.gradle (hunk 4 above)
Comment (rjernst) [inline client/rest/build.gradle]: That is my point. We shade the dependency itself. Nothing in src should be referring to unshaded packages, so there should be no need to do t his relocation here.

**Comment by rjernst** [comment]
Just a note about this in general: I think as a followup (for rc/GA) we need to pull javadocs from the deps and add them to the generated client jar javadocs.

File: core/src/main/resources/org/elasticsearch/bootstrap/test-framework.policy (hunk 3 above)
Comment (hub-cap) [inline core/src/main/resources/org/elasticsearch/bootstrap/test-framework.policy]: I lied, it does, in an IDE /sadface

**Comment by nik9000** [comment]
> also think we should think of a better package name prefix, since these are part of the api

We could, in the grand tradition of Java, go super verbose with something like `org.elasticsearch.client.shaded.httpcomponents`.

File: client/rest/build.gradle [hunk 6]
```diff
@@ -19,6 +19,29 @@
 
 import org.elasticsearch.gradle.precommit.PrecommitTasks
 
+/**
+ * The rest client is a shaded jar. It contains the source of the rest client, as well as all the dependencies,
+ * shaded to the `internal.` package. 3 artifacts come out of this build process.
```
Comment (nik9000) [inline client/rest/build.gradle]: `internal.` is no longer accurate.

File: client/rest/build.gradle [hunk 7]
```diff
@@ -19,6 +19,29 @@
...
+ *    This jar is removed from the output artifacts in cli, and retained in IDE builds.
+ * 2) A jar that contains *only* the `internal.` dependencies. This is a jar that is built before the src is compiled.
+ *    This jar is only used by the rest client so its compilable. There exists a chicken-egg situation where the src
+ *    needs compilation and depends on `internal.` classes, so an intermediary jar needs to exist to satisfy the compile.
+ *    The `deps` classifier is added to this jar. This jar is removed from the output artifacts in cli, and retained in
+ *    IDE builds.
+ * 3) The *actual* jar that will be used by clients. This has no classifier, contains the rest client src and `internal.`.
+ *    This jar is the only actual output artifact of this job.
+ *
+ *  The IDE does not like removing artifacts and changing configurations on the fly, so the bits that make the build
```
Comment (nik9000) [inline client/rest/build.gradle]: s/The IDE/IDEs/ ? I expect you mean IntelliJ and Eclipse.

**Review by tlrx — COMMENTED** [review]
This looks promising. I have compilation errors in my IDE but maybe it's due to Intellij IDEA 2017.1.

File: client/rest/build.gradle [hunk 8]
```diff
@@ -19,6 +19,29 @@
...
 
+/**
+ * The rest client is a shaded jar. It contains the source of the rest client, as well as all the dependencies,
+ * shaded to the `internal.` package. 3 artifacts come out of this build process.
+ *
+ * 1) The original jar contents, using the `internal.` packages, but without the `internal.` code also in the jar.
+ *    The reason this jar was not overwritten was because gradle knows what its inputs and outputs are for a given task
+ *    and we should not alter that by overwriting the original jar. What we do is add a `nodeps` classifier to it instead.
+ *    This jar is removed from the output artifacts in cli, and retained in IDE builds.
+ * 2) A jar that contains *only* the `internal.` dependencies. This is a jar that is built before the src is compiled.
```
Comment (nik9000) [inline client/rest/build.gradle]: I like this comment but I'd rephrase it, maybe moving the number 2 jar to the number 1 jar and explaining that it is also there for IDEs. I think it'd be good also to explain that we use the shaded package names in the source so we still jar up the bits that come out of javac without any post processing.

File: build.gradle [hunk 9]
```diff
@@ -268,7 +268,7 @@ subprojects {
 // the dependency is added.
 gradle.projectsEvaluated {
   allprojects {
-    if (project.path == ':test:framework') {
+    if (project.path == ':test:framework' || project.path == ':client:test') {
       // :test:framework:test cannot run before and after :core:test
```
Comment (nik9000) [inline build.gradle]: Can you add a comment explaining why `:core:test` also cannot run before `:core:test`?

File: client/rest/build.gradle [hunk 10]
```diff
@@ -38,13 +61,66 @@ publishing {
...
+  doLast {
+    shadedSrcDir.mkdir()
+  }
+}
+
+// builds the actual shaded jar that is being distributed
+shadowJar {
+  configurations = [project.configurations.shade]
+  classifier = null
+  relocate 'org.apache', 'org.elasticsearch.client'
```
Comment (nik9000) [inline client/rest/build.gradle]: I think it'd be clearer to do this with the regular jar task because you are just combining the two jar files into one, right? At least, I thought that is what we'd talked about doing.

File: client/rest/build.gradle [hunk 11]
```diff
@@ -38,13 +61,66 @@ publishing {
...
+}
+
+jar {
+  // rename the default jar because we are removing the classifier from the shadowJar, which is a null classifier
```
Comment (nik9000) [inline client/rest/build.gradle]: I'd say something like "move the default jar to another classifier so the jar with the dependencies can have the empty classifier".

File: core/src/main/resources/org/elasticsearch/bootstrap/test-framework.policy (hunk 3 above)
Comment (nik9000) [inline core/src/main/resources/org/elasticsearch/bootstrap/test-framework.policy]: So we need the grant above for gradle and the grant below for the IDEs which don't use the fully shaded dependency. They use the -nodeps jar. I *think* they should be using the shadedDeps jar rather than the original dependencies. That way the packages line up.

File: build.gradle (hunk 9 above)
Comment (hub-cap) [inline build.gradle]: Ill leave that for ryan to answer heh

File: core/src/main/resources/org/elasticsearch/bootstrap/test-framework.policy (hunk 3 above)
Comment (hub-cap) [inline core/src/main/resources/org/elasticsearch/bootstrap/test-framework.policy]: Id love for them to use the shaded jar itself, but we use project dependencies, and intellij is smarter than me when it comes to dealing w/ dependencies. here is what happens to the other projects, and they seem to (only in intellij) depend on the jar task w/ the funky classifier, even though I remove it from the runtime artifacts and remove the extendsFrom compile. So these stay, just for intellij purposes. And they are using the same packages anyway, since the -deps jar and the no classifier actual shaded jar have the same `relocate` bits.

image

File: client/rest/build.gradle [hunk 12]
```diff
@@ -56,6 +132,8 @@ dependencies {
   signature "org.codehaus.mojo.signature:java17:1.0@signature"
 }
 
+dependencyLicenses.dependencies = project.configurations.shade
```
Comment (hub-cap) [inline client/rest/build.gradle]: Ill chat w/ ryan about this, good call for bringing it up

File: client/rest/build.gradle (hunk 10 above)
Comment (hub-cap) [inline client/rest/build.gradle]: You are right, at this point I could combine the deps jar with the src/ in the project. I can follow up with this, as its a better idea than running another "very similar" shadow task. I can follow up with this

**Comment by rjernst** [comment]
> We could, in the grand tradition of Java, go super verbose with something like org.elasticsearch.client.shaded.httpcomponents

Please let's not do that. The package names have real costs (eg Windows file path limits).

File: build.gradle (hunk 9 above)
Comment (rjernst) [inline build.gradle]: I don't understand your statement @nik9000. The task graph cannot have cycles, and a dependency on yourself is still a cycle.

File: client/rest/build.gradle (hunk 12 above)
Comment (rjernst) [inline client/rest/build.gradle]: Good idea

File: build.gradle (hunk 9 above)
Comment (nik9000) [inline build.gradle]: Sorry, `:client:test` cannot run before `:core:test`. Basically, if we are going to leave a comment explaining why `:test:framework` needs this exemption we should probably explain the `:client:test` too.

**Comment by nik9000** [comment]
> Please let's not do that. The package names have real costs (eg Windows file path limits).

Fine by me. I'd certainly prefer it if all the shaded things were grouped together in some obvious way but I'm not going to block anything over it.

**Review by rjernst — APPROVED** [review]
This works great for me from the CLI. I at first had issues testing it with IntelliJ, in 2016.2. I upgraded to 2017.2, but still had issues. I then tried using the "module per source set" option and it "sort of" worked, but I needed to downgrade to 2017.1. So, I think we should add a note in CONTRIBUTING that using the module per source set option is now required, and probably also a note that 2017.1 is recommended (2017.2 was just recently released so I expect the issue should be fixed in a patch release).

File: build.gradle (hunk 9 above)
Comment (rjernst) [inline build.gradle]: Ok I see the confusion. The comment should be updated. It means that if the code below this were to run, a circular dependency would be added between `:test:framework:test` and `:core:test`. Likewise, the new check is added because `:client:test` would have a circular task dep with `:client:rest:test`.

**Comment by hub-cap** [comment]
👍  ill create a few issues off this initial one based on the feedback here

**Comment by javanna** [comment]
thanks @hub-cap ! Was this supposed to go to 5.x too? I see it has the 5.6 label.

**Comment by hub-cap** [comment]
yes i have backported it, ty for keepin an eye out @javanna <3
